lib/req/steps.ex

```diff
@@ -1636,42 +1636,70 @@ defmodule Req.Steps do
       {request, response}
     else
       codecs = compression_algorithms(Req.Response.get_header(response, "content-encoding"))
-      {decompressed_body, unknown_codecs} = decompress_body(codecs, response.body, [])
-      response = put_in(response.body, decompressed_body)
-
-      response =
-        if unknown_codecs == [] do
-          response
-          |> Req.Response.delete_header("content-encoding")
-          |> Req.Response.delete_header("content-length")
-        else
-          Req.Response.put_header(response, "content-encoding", Enum.join(unknown_codecs, ", "))
-        end
 
-      {request, response}
+      case decompress_body(codecs, response.body, []) do
+        %Req.DecompressError{} = exception ->
+          {request, exception}
+
+        {decompressed_body, unknown_codecs} ->
+          response = put_in(response.body, decompressed_body)
+
+          response =
+            if unknown_codecs == [] do
+              response
+              |> Req.Response.delete_header("content-encoding")
+              |> Req.Response.delete_header("content-length")
+            else
+              Req.Response.put_header(
+                response,
+                "content-encoding",
+                Enum.join(unknown_codecs, ", ")
+              )
+            end
+
+          {request, response}
+      end
     end
   end
 
   defp decompress_body([gzip | rest], body, acc) when gzip in ["gzip", "x-gzip"] do
-    decompress_body(rest, :zlib.gunzip(body), acc)
+    try do
+      decompress_body(rest, :zlib.gunzip(body), acc)
+    rescue
+      e in ErlangError ->
+        if e.original == :data_error do
+          %Req.DecompressError{format: :gzip, data: body}
+        else
+          reraise e, __STACKTRACE__
+        end
+    end
   end
 
   defp decompress_body(["br" | rest], body, acc) do
-    if brotli_loaded?() do
-      {:ok, decompressed} = :brotli.decode(body)
+    with true <- brotli_loaded?(),
+         {:ok, decompressed} <- :brotli.decode(body) do
       decompress_body(rest, decompressed, acc)
     else
-      Logger.debug(":brotli library not loaded, skipping brotli decompression")
-      decompress_body(rest, body, ["br" | acc])
+      false ->
+        Logger.debug(":brotli library not loaded, skipping brotli decompression")
+        decompress_body(rest, body, ["br" | acc])
+
+      :error ->
+        %Req.DecompressError{format: :br, data: body}
```

Comment:
with/1 with multiple clauses tends to be hard to maintain because it isn't immediately clear where it originated from. (Right now it is but if this code grows it ain't gonna.)

Could you go with something like this?

```elixir
if brotli_loaded?() do
  case :brotli.decode(body) do
    {:ok, decompressed} ->
      # ...

    :error ->
      # ...

  end
else
  # ...
end
```